app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,539 @@
+          requestPermissions(new String[] {Manifest.permission.WRITE_EXTERNAL_STORAGE}, PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS);
+        }
+
+        return false;
+      }
+    }
+
+    return true; // Control reaches here: Either permission granted at install time, or at the time of request
+  }
+
+  @Override
+  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                         @NonNull int[] grantResults) {
+    switch (requestCode) {
+      case PERMISSION_REQUEST_CODE_COARSE_LOCATION: {
+        if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
+          Log.e(TAG, "Location permission not granted");
+
+          showToast(this, R.string.permission_refused_location, Toast.LENGTH_LONG);
+          closeLocalFileTransferActivity();
+          break;
+        }
+      }
+
+      case PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS: {
+        if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
+          Log.e(TAG, "Storage write permission not granted");
+
+          showToast(this, R.string.permission_refused_storage, Toast.LENGTH_LONG);
+          closeLocalFileTransferActivity();
+          break;
+        }
+      }
+    }
+  }
+
+  private boolean isLocationServicesEnabled() {
+    LocationManager locationManager = (LocationManager) getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
+    boolean gps_enabled = false;
+    boolean network_enabled = false;
+
+    try {
+      gps_enabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
+    } catch(Exception ex) {ex.printStackTrace();}
+
+    try {
+      network_enabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+    } catch(Exception ex) {ex.printStackTrace();}
+
+    return (gps_enabled || network_enabled);
+  }
+
+  private void requestEnableLocationServices() {
+    FragmentTransaction fragmentTransaction = getSupportFragmentManager().beginTransaction();
+    Fragment prev = getSupportFragmentManager().findFragmentByTag(RequestEnableLocationServicesDialog.TAG);
+
+    if(prev == null) {
+      RequestEnableLocationServicesDialog dialogFragment = new RequestEnableLocationServicesDialog();
+      dialogFragment.show(fragmentTransaction, RequestEnableLocationServicesDialog.TAG);
+    }
+  }
+
+  public static class RequestEnableLocationServicesDialog extends DialogFragment {
+    
+    public static final String TAG = "LocationDialog";
+    
+    @NonNull
+    @Override
+    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
+
+      AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
+      builder.setMessage(R.string.request_enable_location)
+          .setPositiveButton(R.string.open_location_settings, new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(DialogInterface paramDialogInterface, int paramInt) {
+              paramDialogInterface.cancel();
```

Comment:
you don't need to cancel, a click action on a dialog implicitly closes it

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/c6de4045c87b493399f574679e8b26614ed15132#diff-e1d59ae6e15dcb502e125fb6e243d15dL447)